Dispatches coroutines in the correct thread on suspending endpoints

pak3nuh opened the pull request:
fixes #16739

When reasteasy reactive endpoints are delegated to the correct bean, it ensures the coroutine is dispatched in the same context that was processing the endpoint.
If the endpoint has not a Vertx context, it launches the coroutine in the IO dispatcher. This was observed when we add the @Blocking annotation a suspending the endpoint.

pak3nuh committed:
Runs continuations on vertx context

Captures the vertx context from the resteasy endpoint to use it in the dispatcher.
If the endpoint doesn't use a vertx context, dispatches through the IO dispatcher.
Adds specific classes for coroutines for future user customization, once exposed.

pak3nuh commented:
@evanchooly @FroMage could you please review?

mschorsch commented:
> fixes #16739
> 
> When reasteasy reactive endpoints are delegated to the correct bean, it ensures the coroutine is dispatched in the same context that was processing the endpoint.
> If the endpoint has not a Vertx context, it launches the coroutine in the IO dispatcher. This was observed when we add the @Blocking annotation a suspending the endpint.

I

> fixes #16739
> 
> When reasteasy reactive endpoints are delegated to the correct bean, it ensures the coroutine is dispatched in the same context that was processing the endpoint.
> If the endpoint has not a Vertx context, it launches the coroutine in the IO dispatcher. This was observed when we add the @Blocking annotation a suspending the endpoint.

I think we should not rely on the Dispatchers.IO thread pool and instead use the quarkus worker pool for @blocking (https://quarkus.io/guides/context-propagation#usage-example-for-completionstage).

pak3nuh committed:
Runs blocking code in the ManagedExecutor

pak3nuh commented:
> > fixes #16739
> > When reasteasy reactive endpoints are delegated to the correct bean, it ensures the coroutine is dispatched in the same context that was processing the endpoint.
> > If the endpoint has not a Vertx context, it launches the coroutine in the IO dispatcher. This was observed when we add the @Blocking annotation a suspending the endpoint.
> 
> Looks great .
> 
> I think we should not rely on the Dispatchers.IO thread pool and instead use the quarkus worker pool for @Blocking (https://quarkus.io/guides/context-propagation#usage-example-for-completionstage).

Nice catch @mschorsch was actually trying to fix that in the first place :sweat_smile:

mschorsch commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/kotlin/src/main/kotlin/org/jboss/resteasy/reactive/server/runtime/kotlin/CoroutineInvocationHandler.kt:
```diff
@@ -1,25 +1,43 @@
+
+        val dispatcher: CoroutineDispatcher = Vertx.currentContext()?.let(::VertxDispatcher)
+                // The @Blocking annotation will not run in a Vertx context
+                ?: executorDispatcher
```
How about [asCoroutineDispatcher](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/java.util.concurrent.-executor/as-coroutine-dispatcher.html)?

```kotlin
        val dispatcher: CoroutineDispatcher = Vertx.currentContext()?.let(::VertxDispatcher)
                // The @Blocking annotation will not run in a Vertx context
                ?: managedExecutor.asCoroutineDispatcher()
```

pak3nuh replied:
The wrapper dispatcher around executor services is not deterministic. If the executor service rejects the task, it gets submitted to `Dispatchers.IO`.
I think is best to keep it deterministic and if the task submission fails, for instance, the executor service is being shut down, then it will fail the request.

mschorsch replied:
Your're right, nice catch 👍.

mschorsch replied:
After reading more about [asCoroutineDispatcher](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/java.util.concurrent.-executor/as-coroutine-dispatcher.html), i'm not sure anymore.

It is correkt, that the task is submitted to the [Dispatchers.IO](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-i-o.html), but only `... to cleanup its resources and promptly complete` ([Code](https://github.com/Kotlin/kotlinx.coroutines/blob/32336be8fa90818cff9f2ba3c36b3f7c89e079c4/kotlinx-coroutines-core/jvm/src/Executors.kt#L92-L100)).

It might be better to use [asCoroutineDispatcher](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/java.util.concurrent.-executor/as-coroutine-dispatcher.html) in this case, because it is an `Executor` :thinking:

pak3nuh replied:
Don't see the advantage of being an `Executor` per se.

Regarding the resource cleanup it makes sense. I'm just wondering If I want to run additional code when the Quarkus runtime already decided to shutdown its executors.
Would like to have feedback from other reviewers on this one.

pak3nuh replied:
Since we didn't had any additional feedback and the latest discussions removed the executor, I'll resolve this as well.

mschorsch commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/kotlin/src/main/kotlin/org/jboss/resteasy/reactive/server/runtime/kotlin/ApplicationCoroutineScope.kt:
```diff
@@ -0,0 +1,42 @@
+ * This should be the main interception point to place user specific [CoroutineContext.Element] objects. Things
+ * like activity ids, logger context propagation, tracing, exception handlers, etc.
+ */
+@ApplicationScoped
```
`@Singleton`?

pak3nuh replied:
I would prefer to keep the ApplicationScope for now.
AFAIK In the context of Quarkus applications, there is only one application scope, technically rendering them as singletons.
If for some reason there is more than one application scope, I wouldn't want them to share the same execution context, because it would break the structured concurrency principle for that application instance.
Do you see any advantage to use `Singleton` in this case?

stuartwdouglas replied:
They are mostly the same, the only difference is @Singleton does not use a Client Proxy, while ApplicationScoped does. This means that application scoped is lazily initialized when it is invoked, while singleton is eagerly initialized when it is injected/looked up for the first time.

ApplicationScoped can also support circular references in injection.

stuartwdouglas replied:
In this case Singleton is probably better, as there is no real need for the client proxy.

pak3nuh replied:
The CDI spec is not very clear about this, but it is as you said, no clear advantage in allowing proxies at this point. Will change.
Thanks for the clarification @stuartwdouglas

mschorsch commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/kotlin/src/main/kotlin/org/jboss/resteasy/reactive/server/runtime/kotlin/ApplicationCoroutineScope.kt:
```diff
@@ -0,0 +1,42 @@
+class ApplicationCoroutineScope : CoroutineScope, AutoCloseable {
+    override val coroutineContext: CoroutineContext = SupervisorJob()
+
+    override fun close() {
```
Why is this class `AutoCloseable`? At which point is `close called?

pak3nuh replied:
The intention was to signal the container that the bean should be closed on application shutdown. Spring does this automatically and I've seen other frameworks do it as well. Doesn't Quarkus do this on the runtime?
On build time I have `QuarkusBuildCloseablesBuildItem` that seems to do that explicitly. Should I do it that way instead?

mschorsch replied:
>  Spring does this automatically and I've seen other frameworks do it as well. Doesn't Quarkus do this on the runtime?

I don't think Quarkus automatically takes `AutoCloseable` into account. There are [lifecycle callbacks](https://quarkus.io/guides/cdi#lifecycle-callbacks) and annotations for [startup and shutdown events](https://quarkus.io/guides/lifecycle).

> On build time I have QuarkusBuildCloseablesBuildItem that seems to do that explicitly. Should I do it that way instead?

Possibly, but I don't know enough about it.

stuartwdouglas replied:
You should put a @PreDestory annotation on the method and it will be called when the app is shutdown. QuarkusBuildCloseablesBuildItem is for things that need to be closed after the application is built, its not a runtime thing.

pak3nuh replied:
Thanks for the clarification. Added the annotation.

mschorsch commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/kotlin/src/main/kotlin/org/jboss/resteasy/reactive/server/runtime/kotlin/ApplicationCoroutineScope.kt:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Dispatches the coroutine in a worker thread from Vertx.
+ */
+class VertxDispatcher(private val vertxContext: Context) : CoroutineDispatcher() {
```
Here is how Vert.x solves the problem:
https://github.com/vert-x3/vertx-lang-kotlin/blob/0af356249b2f63b9abf58ebf40103981d654e173/vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt#L223

pak3nuh replied:
Won't the proposed implementation achieve the same result without pulling another dependency?
Are there other advantages of using this approach?

mschorsch replied:
I've thought it would be better to use a battle tested version of a `CoroutineDispatcher` implementation, but after thinking a little more about it and in the light of your previous post about `asCoroutineDispatcher` i came to the conclusion that your implementation is much better 🚀 .

mschorsch replied:
What do you think of the following optimization? 

If we are on an Vert.x event thread, then we don't need to start the whole `runOnContext` maschinery.

```kotlin
class VertxDispatcher(private val vertxContext: Context) : CoroutineDispatcher() {

    override fun dispatch(context: CoroutineContext, block: Runnable) {
        if (Vertx.currentContext() != vertxContext) {
            vertxContext.runOnContext { block.run() }
        } else {
            block.run()
        }
    }
}
```

pak3nuh replied:
This would violate the rule that we [mustn't call the block immediately on this method](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/common/src/CoroutineDispatcher.kt#L78).

mschorsch replied:
Oh ok. Thanks for the hint.

mschorsch commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/kotlin/src/main/kotlin/org/jboss/resteasy/reactive/server/runtime/kotlin/ApplicationCoroutineScope.kt:
```diff
@@ -0,0 +1,42 @@
+
+class ExecutorDispatcher(private val managedExecutor: ManagedExecutor): CoroutineDispatcher() {
+    override fun dispatch(context: CoroutineContext, block: Runnable) {
+        managedExecutor.runAsync(block)
```
Should we use `managedExecutor.execute(block)`? `runAsync` returns a `CompletableFuture`.

pak3nuh replied:
Makes more sense. Will test and update if needed.

mschorsch replied:
>  This method should generally be exception-safe. An exception thrown from this method
 >  may leave the coroutines that use this dispatcher in the inconsistent and hard to debug state.

https://github.com/Kotlin/kotlinx.coroutines/blob/32336be8fa90818cff9f2ba3c36b3f7c89e079c4/kotlinx-coroutines-core/common/src/CoroutineDispatcher.kt#L70

As mentioned before, I would prefer asCoroutineDispatcher because the implementation of dispatch has subtle quirks, but that's my opinion.

pak3nuh replied:
I understand your point. Will solve this issue because it is mainly a duplicate of the above.

pak3nuh committed:
Small changes based on review inputs

Disposes of coroutine scope explicitly.
Changes the executor continuation submission method.

evanchooly commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/kotlin/pom.xml:
```diff
@@ -30,7 +30,6 @@
         <dependency>
             <groupId>org.jetbrains.kotlinx</groupId>
             <artifactId>kotlinx-coroutines-core-jvm</artifactId>
-            <optional>true</optional>
```
This has to be optional so that it doesn't show up in downstream modules/projects unless specifically asked for.

pak3nuh replied:
You are correct. I assumed the kotlin module would be imported explicitelly, but it makes integral part of the resteasy reactive.
I'll change it to produce the coroutine scope only if the coroutine dependency is present.
I'll also try to add a compile-time error for when the user creates suspending endpoints, but doesn't have the dependency in the POM. Currently, the request just hangs.

pak3nuh replied:
Fixed on this [commit](https://github.com/quarkusio/quarkus/pull/16931/commits/1271ce3d9cdc0a4336e6b998429deced0d92e92f)

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/kotlin/src/main/kotlin/org/jboss/resteasy/reactive/server/runtime/kotlin/CoroutineInvocationHandler.kt:
```diff
@@ -1,25 +1,43 @@
+            return
+        }
+
+        val dispatcher: CoroutineDispatcher = Vertx.currentContext()?.let(::VertxDispatcher)
```
I don't understand how this selects between `@Blocking` and not?

pak3nuh replied:
This one was tested empirically. When I defined `@Blocking` endpoints, the vertx context was null and it was running one of the managed executor threads.
Maybe there is a better way to make the distinction?

geoand replied:
Perhaps we should just disallow the case of `@Blocking` when a suspending function is being used as this use case would not yield any runtime benefit I can think of

pak3nuh replied:
On one side I agree with you @geoand. You don't use reactive when you have blocking code.

On the other side, most applications have some kind of blocking function call. In that case, it needs to be made on thread that can block. It would be very nice to have built-in support for that instead of leaving it to the user.

geoand replied:
My point is more that if we do allow this, users will think they are gaining some kind of runtime benefit, when in fact that are gaining none and instead the lines of blocking vs non-blocking just become more blurry.

pak3nuh replied:
That already happens with the current resteasy reactive extension right?

geoand replied:
Does what already happen?

pak3nuh replied:
The fact that you can add the `@Blocking` annotation to a normal endpoint. Resteasy reactive extension already supports it and runs the code on a different thread.
I was planning to do the same with suspending endpoints to be consistent.

geoand replied:
Yes, that is supported by IMHO, it doesn't make sense for suspended functions. @FroMage WDYT?

stuartwdouglas replied:
IMHO this should always just dispatch back to the vert.x executor. One potential issue is that you can have situations where something earlier in the chain has dispatched to a worker (e.g. Security JPA), so you are not currently on the IO thread even though this is where we want to execute. In this situation we would need to dispatch back I think. 

I think for now we should just be running everything on the IO thread, and if there is demand for @Blocking coroutines we can revisit it, but I don't really think they make sense at this point.

pak3nuh replied:
Ok, so we can state we don't support `@Blocking`, i.e. will have unpredictable behavior.
In this case, can I assume the request needs to come from Vertx?
```suggestion
        val dispatcher: CoroutineDispatcher = Vertx.currentContext()?.let(::VertxDispatcher) ?: throw IllegalStateExcetion()
```
Because currently, this can happen. I'm just not sure if it only happens with `@Blocking` or some other feature.

geoand replied:
We can add simply prevent the use of `@Blocking` with corroutines at build time so you won't need to worry about it at runtime.
This can be done in a follow up

pak3nuh replied:
Agreed.
I've blocked usages of `@Blocking` (pun intended) annotations with suspending functions.
We can add support later.

FroMage replied:
Agreed, I don't see how it can make sense to want a coroutine on a fresh thread (even if it's really a thread from a thread pool): they're meant to be reusable for fast suspend, which imply IO threads. Let's detect and disallow at build time.

geoand replied:
I'll do a follow-up once this is in unless @pak3nuh has something ready to go and wants to update the PR

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/kotlin/src/main/kotlin/org/jboss/resteasy/reactive/server/runtime/kotlin/ApplicationCoroutineScope.kt:
```diff
@@ -0,0 +1,46 @@
+class ExecutorDispatcher(private val managedExecutor: ManagedExecutor): CoroutineDispatcher() {
+    override fun dispatch(context: CoroutineContext, block: Runnable) {
+        // todo may need to handle failed job submissions
+        managedExecutor.execute(block)
```
This will have context propagation.

pak3nuh replied:
Not sure if you mean that as a bad thing. I assume we want context propagation right?

pak3nuh replied:
As discussed with @stuartwdouglas, in order to enable context propagation some changes need to be made first and they fall a bit out of scope of this PR.
Will resolve this conversation.

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/kotlin/src/main/kotlin/org/jboss/resteasy/reactive/server/runtime/kotlin/ApplicationCoroutineScope.kt:
```diff
@@ -0,0 +1,46 @@
+class VertxDispatcher(private val vertxContext: Context) : CoroutineDispatcher() {
+    override fun dispatch(context: CoroutineContext, block: Runnable) {
+        vertxContext.runOnContext {
+            block.run()
```
This will not have context propagation: you need to use `ThreadContext` and capture context at the right place to restore it around `block.run()`.

pak3nuh replied:
Thanks for the advice. Will use it based on the response to the earlier comment about context propagation.

stuartwdouglas replied:
We also need to initialize/propagate the request context when we start the dispatch. Currently the request context is not activated for suspended functions.

stuartwdouglas replied:
This does not have to happen in this PR though, as it is a bit of a different issue, but we can't really fix it until this is in as it would conflict.

pak3nuh replied:
Agreed that enabling that would be out of scope for this PR.
Will leave a comment stating that context propagation will need to be handled later.

pak3nuh committed:
Makes coroutines dependencies optional

Avoids dependency propagation on the kotlinx-coroutines-core-jvm artifact.
Only if the artifact is present, the build will create the supporting structures.
Adds a validation in the method scanner for invalid configuration.

pak3nuh committed:
Changes ApplicationCoroutineScope to Singleton

pak3nuh committed:
Forbids usages of Blocking annotation with suspending endpoints

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/KotlinCoroutineIntegrationProcessor.java:
```diff
@@ -54,6 +83,23 @@ MethodScannerBuildItem scanner() {
 
+            private void ensureNotBlocking(MethodInfo method) {
+                if (method.annotation(BLOCKING_ANNOTATION) != null) {
+                    String format = String.format("Suspendable @Blocking methods aren't supported yet: %s.%s",
```
```suggestion
                    String format = String.format("Suspendable @Blocking methods are not supported yet: %s.%s",
```

FroMage reviewed (approved):
LGTM but I'd prefer a Kotlin expert review this.

The plan is to support context propagation in a further PR?

pak3nuh committed:
Some spelling improvements

Co-authored-by: Stéphane Épardaud <stef@inforealm.org>

pak3nuh commented:
> The plan is to support context propagation in a further PR?

That is correct.

> LGTM but I'd prefer a Kotlin expert review this.

I guess we can ask @evanchooly?

evanchooly commented:
I've been following the thread but it's mostly been about vert.x internals that I'm not really qualified to comment on.  but if we're getting close to the final change now, i'd be happy to look over the kotlin aspects.

geoand commented:
So are we ready to merge?

pak3nuh commented:
> So are we ready to merge?

Was trying to see why it was breaking a test in windows, but I can't reproduce the problem. Doesn't seem related to this change dough.

geoand commented:
I restarted CI, let's see what happens

geoand commented:
I am going to merge this so @stuartwdouglas can work on the improvements he mentioned.

I will also follow up with a PR disallowing `@Blocking` on `suspend` methods

geoand commented:
> I will also follow up with a PR disallowing `@Blocking` on `suspend` methods.

I saw you already did, awesome :)
